File: backend/src/main/java/com/hpms/backend/model/Invoice.java
package com.hpms.backend.model;

import com.hpms.backend.enumCollection.PaymentStatusEnum;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Getter
@Setter
@AllArgsConstructor
@NoArgsConstructor
@Entity
@Table(name = "invoice")
public class Invoice {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Enumerated(EnumType.STRING)
    private PaymentStatusEnum paymentStatus;

    private LocalDateTime paymentFulfilledAt;

    @Column(nullable = false)
    private String name;

    private String description;
    private Double totalSum;

    // Invoice recipient fields
    private String recipientName;
    private String recipientCompanyName;
    private String recipientAddress;
    private String recipientCity;
    private String recipientPostalCode;
    private String recipientCountry;
    private String recipientTaxNumber;
    private String recipientEmail;
    private String recipientPhone;

    private LocalDateTime createdAt;
    private LocalDateTime updatedAt;
    private Boolean active = true;

    @OneToMany(mappedBy = "invoice", fetch = FetchType.LAZY)
    private List<Booking> bookings = new ArrayList<>();

    @ManyToMany
    @JoinTable(
            name = "invoice_serviceModels",
            joinColumns = @JoinColumn(name = "invoice_id"),
            inverseJoinColumns = @JoinColumn(name = "serviceModel_id")
    )
    private List<ServiceModel> serviceModels = new ArrayList<>();

    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
        updatedAt = LocalDateTime.now();
    }

    @PreUpdate
    protected void onUpdate() {
        updatedAt = LocalDateTime.now();
    }
}

File: frontend/src/interfaces/invoice.ts
import type { FormValues } from "../lib/form";
import type { BookingData } from "./booking";
import type { PaymentStatus } from "./enums";
import type { ServiceData } from "./service";

export interface InvoiceData {
	id: number;
	paymentStatus: PaymentStatus;
	name: string;
	description?: string;
	totalSum?: number;
	active: boolean;
	createdAt: Date;
	updatedAt?: Date;
	recipientName?: string;
	recipientCompanyName?: string;
	recipientAddress?: string;
	recipientCity?: string;
	recipientPostalCode?: string;
	recipientCountry?: string;
	recipientTaxNumber?: string;
	recipientEmail?: string;
	recipientPhone?: string;
	bookings: BookingData[];
	serviceModels: ServiceData[];
}

export interface CreateInvoicePayload {
	name: string;
	initialBookingId?: number;
}

export interface UpdateInvoicePayload {
	paymentStatus?: PaymentStatus;
	name?: string;
	description?: string;
	recipientName?: string;
	recipientCompanyName?: string;
	recipientAddress?: string;
	recipientCity?: string;
	recipientPostalCode?: string;
	recipientCountry?: string;
	recipientTaxNumber?: string;
	recipientEmail?: string;
	recipientPhone?: string;
	bookingIds?: number[];
	serviceModelIds?: number[];
}

export interface InvoiceFormData extends FormValues {
	paymentStatus: PaymentStatus;
	name: string;
	description: string;
	recipientName: string;
	recipientCompanyName: string;
	recipientAddress: string;
	recipientCity: string;
	recipientPostalCode: string;
	recipientCountry: string;
	recipientTaxNumber: string;
	recipientEmail: string;
	recipientPhone: string;
	bookingIds: number[];
	serviceModelIds: number[];
}
